### PR TITLE
testdrive: Refactor consistency checks, add leaked shard check

### PR DESCRIPTION
**Pull request by ParkMyCar**
This spun out of https://github.com/MaterializeInc/materialize/pull/25173

### Motivation

This PR refactors the consistency checks in testdrive and changes them to run at the end of every test, instead of after every DDL statement, this speeds up testdrive execution. For example, "Testdrive 1" goes from ~15m to ~8m, "Testdrive 6" goes from ~16m to ~14m, and "🐼🏎️ testdrive" in nightly goes from ~1hour 30m to ~1hour. We could miss some inconsistent state, if a later DDL statement corrects the inconsistency of a previous, but I think is worth it for how much faster testdrive runs.

If there is a scenario which a user wants to assert consistency after a statement they can call the new `$ check-consistency` action. Alternatively we can also run the checks after every statement by specifying the `--consistency-check=statement` CLI arg, maybe in a Nightly test.

Also added is a new `$ check-shard-tombstone` which asserts a provided shard ID has been finalized. I wanted to make this a default part of the consistency checks, but subsources currently always leak their shards, so the check would be pretty low signal.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - N/a Test only changes

### CODE REVIEW

**Comment by ParkMyCar** [comment]
Need to investigate the timeouts

File: src/testdrive/src/action/consistency.rs [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+                state.materialize_internal_http_addr,
+            ))
+            .await
+        })
+        .await
+        .context("querying coordinator")?;
+    if response.status() == StatusCode::NOT_FOUND {
+        tracing::info!("Coordinator consistency check not available");
+        return Ok(());
+    }
```
Comment (jkosh44) [inline src/testdrive/src/action/consistency.rs:102]: When do we expect this to happen?

File: src/testdrive/src/action/consistency.rs [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+async fn check_catalog_state(state: &State) -> Result<(), anyhow::Error> {
+    let maybe_disk_catalog = state
+        .with_catalog_copy(|catalog| catalog.state().clone())
+        .await
+        .map_err(|e| anyhow!("failed to read on-disk catalog state: {e}"))?
+        .map(|catalog| catalog.dump().expect("state must be dumpable"));
+    let Some(disk_catalog) = maybe_disk_catalog else {
+        tracing::warn!("No Catalog state on disk, skipping consistency check");
+        return Ok(());
+    };
```
Comment (jkosh44) [inline src/testdrive/src/action/consistency.rs:123]: When do we expect this to happen?

**Review by jkosh44 — APPROVED** [review]

**Comment by def-** [comment]
This has a bunch of nightly failures (thanks for triggering them!)

File: src/testdrive/src/action/consistency.rs (hunk 1 above)
Comment (ParkMyCar) [reply src/testdrive/src/action/consistency.rs:122]: We don't ever expect this to happen, let me upgrade them to errors and see what happens

File: src/testdrive/src/action/consistency.rs (hunk 2 above)
Comment (ParkMyCar) [reply src/testdrive/src/action/consistency.rs:143]: Same as above, this is never expected so let me upgrade this to an error

File: src/testdrive/src/action/consistency.rs (hunk 1 above)
Comment (jkosh44) [reply src/testdrive/src/action/consistency.rs]: Yeah, it's probably better to upgrade these to errors. That way if someone misconfigures a test and these aren't run then they get an error.

**Pull request merged**